### PR TITLE
Cleanups, TODOs, asserts (part 1)

### PR DESCRIPTION
**Pull request by shipilev**
I have been reading the `openjdk/shenandoah` vs `openjdk/jdk` diff and these are the simple code massages we can do to improve and bullet-proof the code, hopefully without changing the semantics. This would continue with more parts later.

Additional testing:
 - [x] macos-aarch64-server-fastdebug (some pre-existing failures)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Reviewers
 * [William Kemper](https://openjdk.org/census#wkemper) (@earthling-amzn - Committer) ⚠️ Review applies to [4c16e6c6](https://git.openjdk.org/shenandoah/pull/245/files/4c16e6c6913981e27673dd9e9ef66964d3e34cc9)
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - Committer) ⚠️ Review applies to [4c16e6c6](https://git.openjdk.org/shenandoah/pull/245/files/4c16e6c6913981e27673dd9e9ef66964d3e34cc9)
 * [Y. Srinivas Ramakrishna](https://openjdk.org/census#ysr) (@ysramakrishna - Author) ⚠️ Review applies to [c0af0d7e](https://git.openjdk.org/shenandoah/pull/245/files/c0af0d7e04174b2d149c0e5b8c37e3454215ad20)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah.git pull/245/head:pull/245` \
`$ git checkout pull/245`

Update a local copy of the PR: \
`$ git checkout pull/245` \
`$ git pull https://git.openjdk.org/shenandoah.git pull/245/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 245`

View PR using the GUI difftool: \
`$ git pr show -t 245`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah/pull/245.diff">https://git.openjdk.org/shenandoah/pull/245.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/shenandoah/pull/245#issuecomment-1497786273)

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp [hunk 1]
```diff
@@ -901,7 +897,7 @@ bool ShenandoahBarrierSetC2::clone_needs_barrier(Node* src, PhaseGVN& gvn) {
       }
     } else {
       return true;
-    }
+        }
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp:900]: This looks off.

File: src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp [hunk 2]
```diff
@@ -944,6 +940,7 @@ void ShenandoahBarrierSetC2::clone_at_expansion(PhaseMacroExpand* phase, ArrayCo
     Node* gc_state    = phase->transform_later(new LoadBNode(ctrl, mem, gc_state_addr, gc_state_adr_type, TypeInt::BYTE, MemNode::unordered));
     int flags = ShenandoahHeap::HAS_FORWARDED;
     if (ShenandoahIUBarrier) {
+      // TODO: Shouldn't this be OLD_MARKING too?
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp:943]: Generational mode requires SATB barriers right now - IU mode only has the young generation.

File: src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp (hunk 1 above)
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp:900]: Yes, this reverts to "weird' upstream state: 
https://builds.shipilev.net/patch-openjdk-shenandoah/src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp.sdiff.html

No reason to have whitespace changes if that is the only change in the hunk.

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp [hunk 3]
```diff
@@ -115,15 +115,14 @@ bool ShenandoahConcurrentGC::collect(GCCause::Cause cause) {
...
+    // Reset task queue stats here, rather than in mark_concurrent_roots,
     // because remembered set scan will `push` oops into the queues and
     // resetting after this happens will lose those counts.
     TASKQUEUE_STATS_ONLY(_mark.task_queues()->reset_taskqueue_stats());
 
     // Concurrent remembered set scanning
     entry_scan_remembered_set();
-    // When RS scanning yields, we will need a check_cancellation_and_abort()
-    // degeneration point here.
+    // TODO: When RS scanning yields, we will need a check_cancellation_and_abort() degeneration point here.
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:125]: We need to be really careful when transitioning to a degenerated cycle after RS scan is complete. Logically, the RS scan is part of the concurrent root scan, but it didn't get implemented with that pattern.

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp [hunk 4]
```diff
@@ -270,47 +271,45 @@ void ShenandoahControlThread::run_service() {
...
+      switch (_mode) {
+        case concurrent_normal: {
+          // At this point:
+          //  if (generation == YOUNG), this is a normal YOUNG cycle
+          //  if (generation == OLD), this is a bootstrap OLD cycle
+          //  if (generation == GLOBAL), this is a GLOBAL cycle triggered by System.gc()
+          // In all three cases, we want to age old objects if this is an aging cycle
+          if (age_period-- == 0) {
+             heap->set_aging_cycle(true);
+            age_period = ShenandoahAgingCyclePeriod - 1;
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp:284]: Indentation looks inconsistent here

**Review by earthling-amzn — APPROVED** [review]

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp
Comment (kdnilsen) [inline src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp:1]: Almost all changes in this PR are "cosmetic".   Some of the changes in ShenandoahBarrierSet.inline.hpp appear to be beyond cosmetic.  Maybe this should be addressed in a separate PR.

Thinking out loud here: is the new_value argument for write_ref_field_post() only needed for IU-style post-write barriers?

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp (hunk 4 above)
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp]: Fixed! Thanks.

**Review by kdnilsen — APPROVED** [review]

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp (hunk 3 above)
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:125]: Yes, I just converted this to TODO, because it looks like a TODO.

File: src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp (hunk 2 above)
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp]: Ok, removed.

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp:1]: What is non-cosmetic in removing `oop newVal`? It does not seem to be used anywhere. Maybe it would indeed be needed for IU, let's reintroduce it then?

File: src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.cpp [hunk 5]
```diff
@@ -75,21 +75,24 @@ ShenandoahHeuristics::~ShenandoahHeuristics() {
   FREE_C_HEAP_ARRAY(RegionGarbage, _region_data);
 }
 
-size_t ShenandoahHeuristics::select_aged_regions(size_t old_available, size_t num_regions, bool preselected_regions[]) {
+size_t ShenandoahHeuristics::select_aged_regions(size_t old_available, size_t num_regions, bool* preselected_regions) {
   ShenandoahHeap* heap = ShenandoahHeap::heap();
+  if (!heap->mode()->is_generational()) {
+    // TODO: Do we need this check, or assert is enough?
```
[the remark this answers is not in the document]
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.cpp:81]: Good! This should be done separately, removing this TODO then, right? I would be afraid about doing non-conservative changes in cleanup PR like this.

File: src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp]: Right, I removed that in new commit.

File: src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp:900]: Yup, that is deliberate to keep upstream diff cleaner, see https://github.com/openjdk/shenandoah/pull/245#discussion_r1160047465

File: src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp [hunk 6]
```diff
@@ -55,6 +55,7 @@ const double ShenandoahAdaptiveHeuristics::HIGHEST_EXPECTED_AVAILABLE_AT_END = 0
 const double ShenandoahAdaptiveHeuristics::MINIMUM_CONFIDENCE = 0.319; // 25%
 const double ShenandoahAdaptiveHeuristics::MAXIMUM_CONFIDENCE = 3.291; // 99.9%
 
+// TODO: Provide comment here
```
[the remark this answers is not in the document]
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp]: Cool, I refined the TODO with `or remove is no longer needed`.

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp:1]: It should be fine to remove it - the post barrier is only for generational mode. The IU barrier needs the new value in its _pre_ write barrier.

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp (hunk 3 above)
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:126]: @earthling-amzn : should this go into the "entry_mark_)roots()" below then? Clearly, it seems that we can't really have a degeneration point after remembered set scanning from how I understood your comment, so the documentation comment above (to which Aleksey merely added the `TODO:` prefix) is incorrect and should be either removed entirely or corrected suitably?

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentMark.cpp [hunk 7]
```diff
@@ -178,7 +178,8 @@ template<ShenandoahGenerationType GENERATION>
 void ShenandoahMarkConcurrentRootsTask<GENERATION>::work(uint worker_id) {
   ShenandoahConcurrentWorkerSession worker_session(worker_id);
   ShenandoahObjToScanQueue* q = _queue_set->queue(worker_id);
-  ShenandoahObjToScanQueue* old = _old_queue_set == nullptr ? nullptr : _old_queue_set->queue(worker_id);
+  ShenandoahObjToScanQueue* old = (_old_queue_set == nullptr) ?
```
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoahConcurrentMark.cpp:181]: call this `old_q` too like you did above at line 121?

File: src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp [hunk 8]
```diff
@@ -1018,21 +1014,20 @@ void ShenandoahBarrierSetC2::eliminate_gc_barrier(PhaseMacroExpand* macro, Node*
...
-         macro->replace_node(mem, macro->intcon(0));
-         continue;
-       }
-       assert(mem->is_Store(), "store required");
-       macro->replace_node(mem, mem->in(MemNode::Memory));
-     }
+    Node* shift = node->unique_out();
+    Node* addp = shift->unique_out();
+    for (DUIterator_Last jmin, j = addp->last_outs(jmin); j >= jmin; --j) {
+      Node *mem = addp->last_out(j);
```
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp:1020]: `Node* mem`

File: src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp [hunk 9]
```diff
@@ -56,6 +56,7 @@ const double ShenandoahAdaptiveHeuristics::HIGHEST_EXPECTED_AVAILABLE_AT_END = 0
 const double ShenandoahAdaptiveHeuristics::MINIMUM_CONFIDENCE = 0.319; // 25%
 const double ShenandoahAdaptiveHeuristics::MAXIMUM_CONFIDENCE = 3.291; // 99.9%
 
+// TODO: Provide comment here or remove if not used
```
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp:59]: currently used at: https://github.com/openjdk/shenandoah/blob/c0af0d7e04174b2d149c0e5b8c37e3454215ad20/src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp#L468

Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp:59]: Some of these may potentially want to be global tunables; may be...

File: src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.hpp [hunk 10]
```diff
@@ -397,7 +397,7 @@ class ShenandoahHeapRegion {
   bool oop_fill_and_coalesce();
 
   // Like oop_fill_and_coalesce(), but without honoring cancellation requests.
-  bool oop_fill_and_coalesce_wo_cancel();
+  bool oop_fill_and_coalesce_without_cancel();
```
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.hpp:400]: Along similar lines, not sure if you also got `register_object_wo_lock` in this round of cleanups.

**Review by ysramakrishna — APPROVED** [review]
A few comments, but overall looks good; thanks fo the clean-ups (and marking future clean-ups).

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentMark.cpp (hunk 7 above)
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/shenandoahConcurrentMark.cpp:181]: Done.

File: src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp (hunk 8 above)
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp:1020]: Done!

File: src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.hpp (hunk 10 above)
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.hpp:400]: Yup, renamed those as well.

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp (hunk 3 above)
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:126]: I think it's fair to leave it as a TODO. Interrupting he RS scan is something we want and we have tried in the past, but we shelved it because it was more complicated than expected.
